<html>
<head>
<title>pydimer.py</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<style type="text/css">
.s0 { color: #8c8c8c; font-style: italic;}
.s1 { color: #080808;}
.s2 { color: #0033b3;}
.s3 { color: #008080; font-weight: bold;}
.s4 { color: #1750eb;}
</style>
</head>
<body bgcolor="#ffffff">
<table CELLSPACING=0 CELLPADDING=5 COLS=1 WIDTH="100%" BGCOLOR="#c0c0c0" >
<tr><td><center>
<font face="Arial, Helvetica" color="#000000">
pydimer.py</font>
</center></td></tr></table>
<pre><span class="s0">&quot;&quot;&quot; 
Takes FASTA sequences and looks for heterodimer interactions. 
 
Creates output .csv with oligo names, sequences, relevant scores, 
and the best alignment for each pair.   
 
Created on Thu Sep 10 19:31:50 2020 
 
@author: ATyler 
&quot;&quot;&quot;</span>

<span class="s0"># import required modules</span>
<span class="s2">import </span><span class="s1">tkinter</span>
<span class="s2">from </span><span class="s1">tkinter </span><span class="s2">import </span><span class="s1">filedialog</span>
<span class="s2">from </span><span class="s1">Bio </span><span class="s2">import </span><span class="s1">SeqIO</span>
<span class="s2">from </span><span class="s1">Bio </span><span class="s2">import </span><span class="s1">pairwise2</span>
<span class="s2">from </span><span class="s1">Bio.pairwise2 </span><span class="s2">import </span><span class="s1">format_alignment</span>
<span class="s2">import </span><span class="s1">primer3</span>
<span class="s2">import </span><span class="s1">pandas </span><span class="s2">as </span><span class="s1">pd</span>

<span class="s0"># import fasta sequences to compare</span>
<span class="s0"># have user select .fasta file</span>
<span class="s1">root = tkinter.Tk()</span>
<span class="s1">root.withdraw()  </span><span class="s0"># keeps root window from appearing since no GUI</span>
<span class="s1">input_file = filedialog.askopenfilename(</span>
    <span class="s1">parent=root,</span>
    <span class="s1">title=</span><span class="s3">&quot;Select FASTA file to import&quot;</span><span class="s1">,</span>
    <span class="s1">filetypes=[(</span><span class="s3">'FASTA files'</span><span class="s1">, </span><span class="s3">'*.fasta'</span><span class="s1">)])</span>

<span class="s0"># read fasta file into the program</span>
<span class="s0"># generate two sequence lists, the original and a list to remove each from</span>
<span class="s0"># after completing all pairs with one sequence</span>
<span class="s1">static_dict = {rec.id: rec.seq </span><span class="s2">for </span><span class="s1">rec </span><span class="s2">in </span><span class="s1">SeqIO.parse(input_file, </span><span class="s3">&quot;fasta&quot;</span><span class="s1">)}</span>
<span class="s1">shrink_dict = {rec.id: rec.seq </span><span class="s2">for </span><span class="s1">rec </span><span class="s2">in </span><span class="s1">SeqIO.parse(input_file, </span><span class="s3">&quot;fasta&quot;</span><span class="s1">)}</span>
<span class="s0"># generate a table to hold each output type</span>
<span class="s1">Oligo1 = []</span>
<span class="s1">Oligo2 = []</span>
<span class="s1">Seq1 = []</span>
<span class="s1">Seq2 = []</span>
<span class="s1">Tm = []</span>
<span class="s1">dG = []</span>
<span class="s1">align = []</span>
<span class="s0"># sequentially run each sequence in list 1 against all sequences in list 2,</span>
<span class="s0"># generating heterodimer data</span>
<span class="s0"># iterate over every sequence in the static dict, comparing it to all sequences</span>
<span class="s0"># in the &quot;shrink&quot; dict.</span>
<span class="s2">for </span><span class="s1">o1 </span><span class="s2">in </span><span class="s1">static_dict.keys():</span>
    <span class="s2">for </span><span class="s1">o2 </span><span class="s2">in </span><span class="s1">shrink_dict.keys():</span>
        <span class="s0"># ensure the same primer isn't ran against itself</span>
        <span class="s2">if </span><span class="s1">static_dict[o1]._data != shrink_dict[o2]._data:</span>
            <span class="s0"># calc heterodimer information</span>
            <span class="s1">result = primer3.calcHeterodimer(static_dict[o1]._data,</span>
                                             <span class="s1">shrink_dict[o2]._data)</span>
            <span class="s0"># if structure is found, append relevant values to dict</span>
            <span class="s2">if </span><span class="s1">result.structure_found:</span>
                <span class="s1">Oligo1.append(o1)</span>
                <span class="s1">Oligo2.append(o2)</span>
                <span class="s1">Tm.append(result.tm)</span>
                <span class="s1">dG.append(result.dg)</span>
                <span class="s1">Seq1.append(static_dict[o1]._data)</span>
                <span class="s1">Seq2.append(shrink_dict[o2]._data)</span>
                <span class="s0"># generate pairwise alignment with no gaps</span>
                <span class="s1">aln = pairwise2.align.localxs(static_dict[o1]._data,</span>
                                              <span class="s1">shrink_dict[o2].reverse_complement()._data,</span>
                                              <span class="s1">-</span><span class="s4">100</span><span class="s1">, -</span><span class="s4">100</span><span class="s1">)</span>
                <span class="s0"># collect the alignment information and format as string</span>
                <span class="s1">top_aln = aln[</span><span class="s4">0</span><span class="s1">]</span>
                <span class="s1">fin_aln = (format_alignment(*top_aln, full_sequences=</span><span class="s2">True</span><span class="s1">))</span>
                <span class="s0"># change bottom string back to original sequence</span>
                <span class="s0"># convert alignment string to list for iteration</span>
                <span class="s1">fin_aln = list(fin_aln)</span>
                <span class="s0"># provide key for conversion</span>
                <span class="s1">char = {</span><span class="s3">&quot;G&quot;</span><span class="s1">: </span><span class="s3">&quot;C&quot;</span><span class="s1">, </span><span class="s3">&quot;C&quot;</span><span class="s1">: </span><span class="s3">&quot;G&quot;</span><span class="s1">, </span><span class="s3">&quot;A&quot;</span><span class="s1">: </span><span class="s3">&quot;T&quot;</span><span class="s1">, </span><span class="s3">&quot;T&quot;</span><span class="s1">: </span><span class="s3">&quot;A&quot;</span><span class="s1">}</span>
                <span class="s0"># iterate over string, converting complements after the &quot;alignment&quot;</span>
                <span class="s0"># line (which always contains at least one &quot;|&quot;)</span>
                <span class="s1">bot = </span><span class="s2">False</span>
                <span class="s2">for </span><span class="s1">index, item </span><span class="s2">in </span><span class="s1">enumerate(fin_aln):</span>
                    <span class="s2">for </span><span class="s1">key, value </span><span class="s2">in </span><span class="s1">char.items():</span>
                        <span class="s2">if </span><span class="s1">item == key </span><span class="s2">and </span><span class="s1">bot:</span>
                            <span class="s1">fin_aln[index] = value</span>
                        <span class="s2">if </span><span class="s1">item == </span><span class="s3">&quot;|&quot;</span><span class="s1">:</span>
                            <span class="s1">bot = </span><span class="s2">True</span>
                <span class="s0"># join corrected string back together</span>
                <span class="s1">fin_aln = </span><span class="s3">&quot;&quot;</span><span class="s1">.join(fin_aln)</span>
                <span class="s0"># append to the relevant dict for storage</span>
                <span class="s1">align.append(fin_aln)</span>
    <span class="s0"># when each sequence has been tested against all remaining sequences,</span>
    <span class="s0"># remove from list 2 so that pairs aren't duplicated</span>
    <span class="s1">shrink_dict.pop(o1)</span>
<span class="s0"># store each set of sequences and their heterodimer data in the matrix</span>

<span class="s1">data = {</span><span class="s3">'Oligo1'</span><span class="s1">: Oligo1, </span><span class="s3">'Oligo2'</span><span class="s1">: Oligo2, </span><span class="s3">'Seq1'</span><span class="s1">: Seq1, </span><span class="s3">'Seq2'</span><span class="s1">: Seq2, </span><span class="s3">'Tm'</span><span class="s1">: Tm, </span><span class="s3">'dG'</span><span class="s1">: dG, </span><span class="s3">'Alignment'</span><span class="s1">: align}</span>
<span class="s1">df = pd.DataFrame(data)</span>

<span class="s0"># export as CSV</span>
<span class="s1">export_file_path = filedialog.asksaveasfilename(defaultextension=</span><span class="s3">'.csv'</span><span class="s1">,</span>
                                                <span class="s1">title=</span><span class="s3">&quot;Choose filename for results&quot;</span><span class="s1">)</span>
<span class="s1">df.to_csv(export_file_path, index=</span><span class="s2">False</span><span class="s1">, header=</span><span class="s2">True</span><span class="s1">)</span>
</pre>
</body>
</html>